### DISCUSSION TITLE
Proportional Nerd Font

### DISCUSSION THREAD

**Discussion by Finii** [ryanoasis/nerd-fonts#951] [General]
This should be the collector for thoughts about the variants we pre-patch.

There are basically (at least) 3 possible variants, while we historically only provide 2 of them. We multiply these two by 2 to get 'normal' and 'Windows Compat' fonts.

`Nerd Font Mono`
`Nerd Font`
`Nerd Font Mono Windows Compat`
`Nerd Font Windows Compat`

_versus_

`Nerd Font Mono`
`Nerd Font`
`Nerd Font Propo`

:-}

Some links:
* https://github.com/ryanoasis/nerd-fonts/issues/442#issuecomment-1263358904
* https://github.com/ryanoasis/nerd-fonts/issues/746#issuecomment-1003694798
* https://github.com/ryanoasis/nerd-fonts/issues/746#issuecomment-1003734095
* https://github.com/ryanoasis/nerd-fonts/issues/731#issuecomment-1003071050
* https://github.com/ryanoasis/nerd-fonts/issues/898#issuecomment-1230144415

Some other links:
* https://github.com/ryanoasis/nerd-fonts/issues/919

**Comment by patrick96** [comment]
*Posting this here to not spam #442*

A proportional font was proposed as a solution to the issue of icons extending into the next cell(s) and overlapping with text in the non-mono fonts in GUI environments that don't explicitly add in extra spaces after icons (in my case [polybar](https://github.com/polybar/polybar)).

I want to give a caveat as to why, in polybar at least, that may not be the optimal solution.
If icons have different cell widths and an icon is changed to another icon, in an animation or due to a state change (e.g. volume change), all subsequent text (or preceding if it is right-aligned) will suddenly be moved a bit to the left or the right.
This can be quite visually jarring and many of our users (myself included) have gone to great lengths to make sure no such movement happens.

The optimal solution, in my opinion, would be for all icons to have the same cell width, but still be larger than regular text (which is what `Symbols Nerd Font Mono` seems to do).
I'm not asking you to pre-patch yet another variant with these properties, I don't think there would be many other use-cases for this.
Conceptually, this would simply be the best of both worlds, in whatever way it is achieved in the end.
Instructing our users to use a regular text font and add a fallback to `Symbols Nerd Font Mono` seems a reasonable way to achieve the same goal.
Font fallback and precise control over which font is used for which characters has been in polybar from the beginning.

From an ease-of-use standpoint a proportional font is definitely preferable and I will instruct our users to use the proportional variant if it is shipped in a release, of course with the above mentioned caveat.

**Reply by Finii** [reply]
Thank you @patrick96 for your thoughts.

Regarding the cell-width-change on related icons. Recently we added a method where 'groups of related icons' can be defined. They are then handled all the same (i.e. scaling and aligning). The scaling and shifting is done on the combined boundingbox of all glyohs in that group.

Driver for this were the moon phases icons, if scaled individually to maximum size they indeed become unusable in a way similar to what you describe:

![image](https://user-images.githubusercontent.com/16012374/196034268-38901589-f24a-440a-8cc7-373606842dfa.png)

In the example above all icons are scaled to maximum size within a cell; which means `width` here. Because the crescent moon is so small it will be scaled up much more than the full moon...

Or this:
![image](https://user-images.githubusercontent.com/16012374/196034880-95d88169-0d23-4c34-8929-b3d9574c5b06.png)

This will be fixed by
* #916 

The example above is the `Nerd Font Mono` variant. What we can easily add then is to give all icons in a group the same advance width also in the proportional font. At the moment the proportional fonts retain the symbol-source's advance width, which may or may not be useful. If I see your 'complains', I guess it is not useful enough.

Ah yes, as you say, the advance width is different in the proportional font, because it is kept from the source:
![image](https://user-images.githubusercontent.com/16012374/196034954-91bc2dfc-124c-4634-8aaf-c6b037b9f8de.png)

Solved for mono here https://github.com/ryanoasis/nerd-fonts/pull/916/commits/16da0f820ea43964b89204065fa79639c33deb9f

Of course that needs manual work and defining all these groups.

Regarding the `Nerd Font Proportional` font variant, maybe we should just try that out with one font face or another. Popular font. I have no idea which fonts are popular and which not (I can not access the github statistics, and anyhow they are of limited used because of the third party packaging via AUR, Cask, ...).
I have the impression JetBrains Mono is used often in GUI-ish contexts, if I remember the issues here correctly.
What do you think?
